### PR TITLE
Acid smoke now checks for Acid protection instead of bio protection.

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## About The Pull Request
Makes smoke_contact check for acid soft and hard protection whenever dealing acid damage instead of permeability 
Optimizes get_permeability_protection

## Why It's Good For The Game
Biological protection should apply to biological sources , acid protection to acid sources.

## Changelog
:cl:
code: Optimized get_permeability_protection()
balance: Xeno smoke now checks for acid protection(both soft and hard) whenever dealing burn damage (from XENO_SMOKE_ACID sources) instead of bio.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/living_defense.dm [hunk 1]
```diff
@@ -257,27 +257,28 @@
...
 	if(CHECK_BITFIELD(S.smoke_traits, SMOKE_BLISTERING))
-		adjustFireLoss(15 * protection)
+		adjustFireLoss(15 * bioProtection)
 		to_chat(src, span_danger("It feels as if you've been dumped into an open fire!"))
 	if(CHECK_BITFIELD(S.smoke_traits, SMOKE_XENO_ACID))
-		if(prob(25 * protection))
+		if(prob(25 * acidProtection))
 			to_chat(src, span_danger("Your skin feels like it is melting away!"))
-		adjustFireLoss(S.strength * rand(20, 23) * protection)
+		adjustFireLoss(max(S.strength * rand(20, 23) * acidProtection - acidHardProtection, 0))
```
Comment (DeltaFire15) [inline code/modules/mob/living/living_defense.dm:271]: Hard armor is not flat DR since a long time. It is anti-AP.

**Review by DeltaFire15 — COMMENTED** [review]
This is a "stealth" mimir nerf since you are moving smoke effects from the variable it negates onto other defense categories.

Should I ask why this PR is phrased in this particular way, instead of even mentioning mimir anywhere in the description, despite you editing the descriptions having clearly considered this effect?

**Comment by MLGTASTICa** [comment]
> This is a "stealth" mimir nerf since you are moving smoke effects from the variable it negates onto other defense categories.
> 
> Should I ask why this PR is phrased in this particular way, instead of even mentioning mimir anywhere in the PR body, despite you editing the item descriptions having clearly considered this effect?

image

File: code/modules/mob/living/living_defense.dm [hunk 2]
```diff
@@ -257,27 +257,28 @@
 	smoke_contact(S)
 
 /mob/living/proc/smoke_contact(obj/effect/particle_effect/smoke/S)
-	var/protection = max(1 - get_permeability_protection() * S.bio_protection, 0)
+	var/bioProtection = max(1 - get_permeability_protection() * S.bio_protection, 0)
+	var/acidProtection = max(1 - get_soft_acid_protection(), 0)
+	var/acidHardProtection = get_hard_acid_protection()
```
Comment (TiviPlus) [inline code/modules/mob/living/living_defense.dm:262]: use snake_case

File: code/modules/mob/living/living.dm [hunk 3]
```diff
@@ -463,6 +463,12 @@
 /mob/living/proc/get_permeability_protection()
 	return LIVING_PERM_COEFF
 
+/mob/living/proc/get_soft_acid_protection()
+	return soft_armor?.getRating(ACID)
+
+/mob/living/proc/get_hard_acid_protection()
+	return hard_armor?.getRating(ACID)
+
```
Comment (TiviPlus) [inline code/modules/mob/living/living.dm:471]: autodoc

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/living_defense.dm (hunk 1 above)
Comment (MLGTASTICa) [reply code/modules/mob/living/living_defense.dm]: Acid gas has no AP variable so this seems like a reasonable work-around.

**Review by Kuro020 — APPROVED** [review]

File: code/modules/mob/living/living_defense.dm (hunk 1 above)
Comment (Lumipharon) [reply code/modules/mob/living/living_defense.dm]: ?

File: code/modules/mob/living/carbon/human/human_helpers.dm [hunk 4]
```diff
@@ -215,25 +215,40 @@
...
+	var/protection = 0
+	for(var/defZone in GLOB.human_body_parts)
+		protection += get_soft_armor(ACID, defZone)
+	// adds arms and feet twice since precise.
+	return protection/11
+
+/mob/living/carbon/human/get_hard_acid_protection()
+	var/protection = 0
+	for(var/defZone in GLOB.human_body_parts)
+		protection += get_hard_armor(ACID, defZone)
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/modules/mob/living/carbon/human/human_helpers.dm:248]:
```suggestion
	for(var/def_zone in GLOB.human_body_parts)
		protection += get_hard_armor(ACID, def_zone)
```

**Comment by CVSthePharmacy** [comment]
> > This is a "stealth" mimir nerf since you are moving smoke effects from the variable it negates onto other defense categories.
> > Should I ask why this PR is phrased in this particular way, instead of even mentioning mimir anywhere in the PR body, despite you editing the item descriptions having clearly considered this effect?
> 
> image I rephrased the description so the change is visible on the mimir , if i wanted this to be a "stealth" change i would've never rewritten the mimir description to reflect the change. Not to mention this BUFFS marine protection in general against acid. Normal armor only had 20% , now since it takes acid into account , the minimum is 35% up to 55%.

i don't believe you, also you didn't put it in the PR body, that's why we think of it as a stealth nerf. Adding onto this, acid protection should be applied on TOP of bio protection, as acid should only affect you if it can get through your armor.

**Comment by lbnesquik** [comment]
That might have broken gas in general.

**Pull request merged**